Provide better workbench status information

jeff-phillips-18 opened the pull request:
Closes [RHOAIENG-17784](https://issues.redhat.com/browse/RHOAIENG-17784)

## Description
Updates to the workbench table:
  - Items in the table are rearranged
  - Show latest status information in the table
  - Removing the status popover to take the user directly to the event log modal

Adjustments to the status modal
  -Update the title
  -Rearrange items for clarity
  -Replace the progress bar with a progress stepper
  - Placing the progress bar in a tab with an alert
  - Add Edit workbench button and update placement

Enable editing of workbenches during start up from the modal and in kebab menu in the table
  - Content changes to the "Unsaved work" alert

## Screen shots
![image](https://github.com/user-attachments/assets/40e4fc2b-aa8f-4640-b3bf-c2b79d4ec6e1)


![image](https://github.com/user-attachments/assets/c7ead6ef-a716-48d2-9570-d9ba1e4926dd)


![image](https://github.com/user-attachments/assets/4dcaaa54-5691-4e6e-bbc8-a21885af6489)


![image](https://github.com/user-attachments/assets/37250de5-a031-431d-a94b-8d818954a411)


![image](https://github.com/user-attachments/assets/68e909bd-6373-4c51-87f2-dcd9e08661eb)


![image](https://github.com/user-attachments/assets/0b56dbf8-bd0c-4315-8206-0cc3f57bb4eb)


![image](https://github.com/user-attachments/assets/71d672b0-8f2d-43f7-8451-b555f1bf721c)


![image](https://github.com/user-attachments/assets/8cf3acb6-1948-4a13-b111-3dabe740c288)


![image](https://github.com/user-attachments/assets/58444d48-eda3-4491-8ecf-f3921e4c42c1)


![image](https://github.com/user-attachments/assets/066efba8-27b4-4254-8b23-2533efe2923b)


![image](https://github.com/user-attachments/assets/9ee37ddf-b5d7-4c9d-aff2-ffc6a1135288)


![image](https://github.com/user-attachments/assets/e8038eb0-d2de-4b00-a501-7280cb463306)


## How Has This Been Tested?
- Navigate to `Data Science Projects` and select a project with multiple workbenches
- Select the `Workbenches` tab
  - Verify each workbench row has a external link icon next to the name and it is enabled only for running workbenches
  - Verify the status column shows the current status and failures show a reason
  - Verify the kebab menu shows `Edit workbench` enabled for all workbenches
- For a Failed workbench, click on the `Failed` label or text link showing reason for failure
  - Verify the workbench status dialog is shown
  - Verify the status shows Failed and the reason is shown
  - Verify the Progress steps show the completed steps and the failure with any incomplete steps shown following
- Select the `Events log` tab
  - Verify the events are shown
- Stop a running workbench
  - Verify the row shows `Stopping`
- Start a workbench
 - Verify the label show starting
- For the starting workbench, click the label or message text
  - Verify the workbench status dialog is shown
  - Verify the current status is shown and Progress steps show the current completed steps
  - Verify the steps update through the startup process
  - Verify when the final status is shown correctly when complete
- In the Workbench status dialog, click the `Edit workbench` button
  - Verify the spawner page is shown for the workbench
  - Verify there is an info alert at the bottom of the page showing: `Updating this workbench will cause it to restart. If you have any unsaved work, save it to avoid losing your data.`
- For a failed workbench, select the Failed label to bring up the Workbench status dialog
- Click the `Stop workbench` button
  - Verify the workbench is stopped 

## Test Impact
Added unit tests for the workbench status dialog

## Request review criteria:
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change.

/cc @tobiastal @jgiardino @kywalker-rh

tobiastal commented:
Looks great!
Few things:
1. For Running workbenches, the workbench name should use the link (blue) color, matching the external link icon.
2. In the latest state below the status, could you display the status as it’s mentioned in the modal/progress stepper rather than the alert? For example, "_Failed to scale up_" instead of "_pod didn't trigger scale up: 1 Insufficient.."
3. In the modal, can you show the alert only for Failed status? All other statuses don't need this (info) alert

jeff-phillips-18 commented:
> * For Running workbenches, the workbench name should use the link (blue) color, matching the external link icon.
> * In the latest state below the status, could you display the status as it’s mentioned in the modal/progress stepper rather than the alert? For example, "_Failed to scale up_" instead of "_pod didn't trigger scale up: 1 Insufficient.."
> * In the modal, can you show the alert only for Failed status? All other statuses don't need this (info) alert

Fixed these. Updated screen shots above.

tobiastal commented:
Thanks, LGTM!

jenny-s51 reviewed (commented):
I’ve tested the functionality based on the provided instructions, and the updates to the table and status information look great @jeff-phillips-18 .

Quick question regarding the "Workbench status" modal: do we have designs for the "Stopped" state? The [mockups](https://www.figma.com/design/8evRkOSYOOYmHjKXNkFJ2a/Provide-better-workbench-status-information?node-id=36-12456&t=VZWnKmYOsEGKm0Z3-0) show designs for the "Starting" and "Failed" states, but there doesn’t seem to be a design for the "Stopped" state.

For reference, before the workbench is stopped, the modal for a failed state looks like this (which seems correct):

<img width="300" alt="Screenshot 2025-01-28 at 4 49 33 PM" src="https://github.com/user-attachments/assets/47eaf271-5e21-43d8-8b1a-c2749b436a05" />

Once the workbench is stopped via the modal, there's a different error message displayed along with the "Creating resources..." text. Should we also disable the "Stop workbench" button in this case? 
<img width="300" alt="Screenshot 2025-01-28 at 4 49 49 PM" src="https://github.com/user-attachments/assets/69619743-5c7a-45fa-81d6-3e2122d108e2" />

Here’s how the modal looks in the "Running" state (expected):
<img width="300" alt="Screenshot 2025-01-28 at 4 50 40 PM" src="https://github.com/user-attachments/assets/83b10366-1565-45f1-8843-8773c74168f7" />

And after being stopped, the "Creating resources..." text is also shown here:
<img width="300" alt="Screenshot 2025-01-28 at 4 51 11 PM" src="https://github.com/user-attachments/assets/2e48a676-6970-46af-abe7-308f31f176d6" />

tobiastal commented:
Great catch @jenny-s51 !

For the Stopped/Stopping state (as shown in the image), I think we should display only the status (Stopped/Stopping) without the additional state text or alert, and show all steps as not started yet.

Also, instead of disabling the "Stop workbench" button for the Stopped status, we could replace it with a "Start workbench" action instead.

<img width="446" alt="image" src="https://github.com/user-attachments/assets/51d44d51-84fb-4c32-b945-d86e95b10222" />

Griffin-Sullivan commented:
Does editing a failed workbench restart it for you? The warning at the bottom suggests it will autorestart but it doesn't seem to. I'll admit I didn't wait super long before restarting it myself.

Otherwise everything I tested is working as expected. Will retry once new changes are ready.

jeff-phillips-18 commented:
> Does editing a failed workbench restart it for you?

It does work for me but it took~30 seconds for any indiction to occur.

jeff-phillips-18 commented:
@jenny-s51 @tobiastal  I have updated such the the progress can be shown for any state. I included screenshots for the new `Running` and `Stopped` states above (they are in dark theme).

jenny-s51 commented on frontend/src/concepts/notebooks/StartNotebookModal.tsx:
```diff
@@ -0,0 +1,312 @@
+  Tabs,
+  TabTitleText,
+} from '@patternfly/react-core';
+import { Modal, ModalVariant } from '@patternfly/react-core/deprecated';
```
Is there a reason we import the deprecated Modal here over the v6 modal?

jeff-phillips-18 replied:
The modal was already built with the deprecated version. I didn't update as there are lots of instances still using the deprecated version. I have updated it now.

jenny-s51 commented on frontend/src/concepts/notebooks/StartNotebookModal.tsx:
```diff
@@ -0,0 +1,312 @@
+  Flex,
+  FlexItem,
+  List,
+  ListItem,
```
Can import `Modal` and `ModalHeader` from v6

jenny-s51 commented on frontend/src/concepts/notebooks/StartNotebookModal.tsx:
```diff
@@ -0,0 +1,312 @@
+      onClose={() => onClose(false)}
+      showClose={showClose}
+      data-testid="notebook-status-modal"
+      title={
```
Can we delete this prop here and add it to `ModalHeader`

jenny-s51 commented on frontend/src/concepts/notebooks/StartNotebookModal.tsx:
```diff
@@ -0,0 +1,312 @@
+      variant={ModalVariant.small}
+      isOpen
+      onClose={() => onClose(false)}
+      showClose={showClose}
```
Can delete `showClose` if we update to v6 modal since it's set by `onClose`

jenny-s51 commented on frontend/src/concepts/notebooks/StartNotebookModal.tsx:
```diff
@@ -0,0 +1,312 @@
+  return (
+    <Modal
+      appendTo={document.body}
+      variant={ModalVariant.small}
```
`variant="small"`

jeff-phillips-18 replied:
Using the `ModalVariant` enum is better, I left that alone.

jenny-s51 reviewed (commented):
The latest updates look great @jeff-phillips-18 . Tested the functionality and everything works as expected. 

I noticed a couple of padding inconsistencies - the modal content isn’t aligned with the modal title, and there's some extra padding under the action buttons. Left comments below with the suggested fixes (which includes upgrading to the v6 Modal). 

Otherwise LGTM!

Current:
<img width="300" alt="Screenshot 2025-01-30 at 2 19 01 PM" src="https://github.com/user-attachments/assets/8609b247-994c-40dc-b577-143f98f08f32" />


Expected:
<img width="300" alt="Screenshot 2025-01-30 at 2 15 34 PM" src="https://github.com/user-attachments/assets/a40bb683-29e9-4488-9504-3da3f8e2f0d3" />

jeff-phillips-18 commented:
@jenny-s51 Updated to use `Modal` from `react-core`.

jenny-s51 reviewed (commented):
The latest updates to the modal look great, thank you for making these changes @jeff-phillips-18 .

Just one last small adjustment based on @tobiastal 's feedback - could we hide both the error text and the alert for the "Stopped" states? Currently the text and alert are still visible in the modals for workbenches that are both failed and stopped.

<img width="200" alt="Screenshot 2025-01-31 at 10 58 14 AM" src="https://github.com/user-attachments/assets/22deb921-92cd-405c-80f0-1d72a46946d2" />

Griffin-Sullivan commented:
I've tested with the updates and everything LGTM, pending @jenny-s51's suggestion.

jeff-phillips-18 commented:
> Currently the text and alert are still visible in the modals for workbenches that are both failed and stopped.

@jenny-s51 Updated

jenny-s51 reviewed (approved):
/lgtm

christianvogt commented:
Just did some quick tests. Overall it works well.

I noticed some minor issues that can be addressed afterwards:

Sometimes whenever all steps are green, there is still progress being reported:
![image](https://github.com/user-attachments/assets/5f3db215-9a15-4b9e-bf00-ef3f604e6b53)

If I stop the workbench from the modal, it briefly shows that "Server requested" is green with `creating resources` message. This was also mentioned in an earlier comment by @jenny-s51:
![image](https://github.com/user-attachments/assets/dbafe854-edef-4301-8973-5e8a273601d8)
